src/compiler/checker.ts

```diff
@@ -6339,7 +6339,7 @@ namespace ts {
                         const resolvedSymbol = resolveName(param, paramSymbol.name, SymbolFlags.Value, undefined, undefined);
                         paramSymbol = resolvedSymbol;
                     }
-                    if (i === 0 && paramSymbol.name === "this") {
+                    if (i === 0 && paramSymbol.name === "this" || (param.type && param.type.kind === SyntaxKind.JSDocThisType)) {

```

Comment:
I'm not a fan of `JSDocThisType`. It looks like you would have a parameter with no name, and a type of kind `JSDocThisType`, where the type itself has a `.type` property storing the real type.
I think `this: T` should be parsed to the same AST whether we're in TypeScript or in JSDoc, so you don't need a condition for both here.